kv_connectors/llmd_fs_backend/src/csrc/storage/file_io.cpp

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2025 The llm-d Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <filesystem>
+#include <fstream>
+#include <vector>
+#include <cstring>
+#include <cerrno>
+#include <fcntl.h>
+#include <sys/stat.h>
+
+#include "file_io.hpp"
+#include "buffer.hpp"
+
+namespace fs = std::filesystem;
+
+// Thread-local index for CUDA streams
+extern thread_local size_t thread_stream_idx;
+
+// Write a tensor to disk using a temporary file and atomic rename
+bool write_file_to_disk(const std::string& target_path, const torch::Tensor& host_buf) {
+    // Pointer and size of data to write
+    const void* data_ptr = host_buf.data_ptr();
+    size_t nbytes = host_buf.nbytes();
+
+    // Create parent directory if needed
+    fs::path file_path(target_path);
+    fs::path parent_dir = file_path.parent_path();
+    try {
+        fs::create_directories(parent_dir);
+    } catch (const fs::filesystem_error& e) {
+        std::cerr << "[ERROR] Failed to create directories: " << e.what() << "\n";
+        return false;
+    }
+
+    // Write to a temporary file to ensure atomic replace on rename
+    // Include thread_stream_idx so each thread uses a unique temporary file
+    std::string tmp_path = target_path + std::to_string(thread_stream_idx) + ".tmp.";
+
+    // Define a larger buffer (1MB) to reduce syscall overhead and speed up I/O
+    const size_t WRITE_BUFFER_SIZE = 1 * 1024 * 1024;  // 1MB buffer
+
+    std::ofstream ofs(tmp_path, std::ios::out | std::ios::binary);
+    if (!ofs) {
+        std::cerr << "[ERROR] Failed to open temporary file for writing: " << tmp_path << " - " << std::strerror(errno) << "\n";
+        return false;
+    }
+
+    // Allocate custom I/O buffer for this stream (replaces small default buffer)
+    std::vector<char> buffer(WRITE_BUFFER_SIZE);
```

Comment:
Can you try to avoid buffer allocation each call and instead use a thread-local pre-allocated buffer?
e.g.
```
const size_t WRITE_BUFFER_SIZE = 1 * 1024 * 1024; // 1MB buffer
thread_local std::vector<char> thread_write_buffer(WRITE_BUFFER_SIZE);
```